service/src/java/org/apache/hive/service/auth/PlainSaslHelper.java

```diff
@@ -75,6 +89,70 @@ private PlainSaslHelper() {
     throw new UnsupportedOperationException("Can't initialize class");
   }
 
+  public static final class DualSaslTransportFactory extends TTransportFactory {
+    TTransportFactory otherFactory;
+    TTransportFactory noAuthFactory;
+    String hiveHost;
+
+    public DualSaslTransportFactory(TTransportFactory otherFactory, String hiveHost)
+            throws LoginException {
+      this.noAuthFactory = getPlainTransportFactory(AuthMethods.NONE.toString());
+      this.otherFactory = otherFactory;
+      this.hiveHost = hiveHost;
+    }
+
+    // Return true if the IP address is from the same domain as the server.
+    private boolean isSameDomainConnection(InetAddress remoteInetAddress) {
+      String[] hiveHostSplit = null;
+      if (hiveHost != null) {
+        hiveHostSplit = hiveHost.split("\\.");
+      }
+
+      String clientHostName = remoteInetAddress.getCanonicalHostName();
+      String[] clientHostSplit = null;
+      if (clientHostName != null) {
+        clientHostSplit = clientHostName.split("\\.");
+      }
+
+      // TODO: the callers should also pass the server address or obtain in by using some other
+      //  method. Then check whether the connection's remote address is from the same domain as
+      //  the server. For now, for the sake of testing, return true. We need to understand what
+      //  exactly it means to have connection from the same domain.
+      // For now, two hosts are in the same domain if the last two parts in their hostname are same.
+      if (hiveHostSplit.length >= 2 && clientHostSplit.length >= 2) {
```

Comment:
We should check all parts of the hiveHostSplit. So, if hiveHostSplit (provided via config) is something like x.example.com, we should ensure that (only) everything below x.example.com (i.e. das.x.example.com) qualifies instead of checking just the top two levels (which are probably always cloudera and com at the final DWX deployment).

Reply:
As per the latest approach, the trusted domain will be configured and thus we do not rely on hivehost to know the trusted domain.